### PR TITLE
Conversation scaffolding

### PR DESCRIPTION
**Pull request by neekolas**
## Summary

This unblocks conversation creation and messaging. It works around the limitations around passing `Client` instances into async functions by making the builder return an `Arc<Mutex<Client<...>>>`. The Arc is needed to store the Client on the heap, and the mutex to allow for multiple callers to get mutable references to the Client (since the `XmtpApiClient`, `Persistence`, and `MessageStore` require mutability).

It also includes a _very_ rough implementation of initiating a conversation.

## Known shortcomings
- This is clunky and awkward to work with
- This removes our ability to do concurrent operations using clients, since the mutex will block until other callers have finished their work. Effectively makes the app synchronous, even if it is using async functions.
- I can't figure out how to pass the client into an `async_trait`. I previously had a `Conversation` trait, which `SecretConversation` was just one implementation of, but that doesn't seem to work. Probably fixable, and not an urgent issue.

## Next steps
- Rename `ContactBundle` to `InstallationContactBundle` in the protos for clarity
- Define invite message types and use those to initiate conversations
- Database for storing session state
- Store all sessions on the Conversation model in some sort of mapping

### CODE REVIEW

File: xmtp/src/session.rs [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+        Self { session }
+    }
+
+    pub fn id(&self) -> String {
+        self.session.session_id()
+    }
+
+    // TODO: Replace the OlmMessage with our own message wrapper? Or leave up to the caller?
+    pub fn encrypt(&mut self, plaintext: &[u8]) -> OlmMessage {
+        self.session.encrypt(plaintext)
```
Comment (jazzz) [inline xmtp/src/session.rs:18]: This doesn't look like it provides any protection for unknown key-share attacks.

https://gitlab.matrix.org/matrix-org/olm/blob/master/docs/olm.md#message-authentication-concerns

Doesn't necessarily need to be solved here, however there should be a clear place where PreKey messages are validated

**Review by jazzz — APPROVED** [review]
Good work around for now. Hopefully buys some more time before revisiting call model is needed.

File: xmtp/src/session.rs (hunk 1 above)
Comment (neekolas) [reply xmtp/src/session.rs:18]: Once we have invite proto messages, there's a lot more to validate. That's coming up next.

File: xmtp/src/lib.rs [hunk 2]
```diff
@@ -1,11 +1,14 @@
 pub mod account;
 pub mod association;
 pub mod builder;
-mod client;
+pub mod client;
```
[the remark this answers is not in the document]
Comment (neekolas) [reply xmtp/src/lib.rs:4]: Even if we don't want Client to ultimately be public, I kinda assume callers will need access to ClientError

[the remark this answers is not in the document]
Comment (jazzz) [reply xmtp/src/lib.rs:4]: > kinda figure we'll need the Client public
Agreed. Though That doesn't necessarily mean the module needs to be public. 

Pub use statements allow the important components to be reexported with more usable paths:
https://github.com/xmtp/libxmtp/blob/4c5bbc565ed34ce28680150a295339bfd9c775bb/xmtp/src/lib.rs#L15-L16

Example: https://github.com/xmtp/libxmtp/blob/4c5bbc565ed34ce28680150a295339bfd9c775bb/xmtp/examples/cli-client.rs#L24

While the number of consumers is low I don't think it matters, however as it grows we'll want to re-exporting at the lib/ mod levels to break less dependencies during rafactors

**Pull request merged**